autoscaler/kubernetes_client.go

```diff
@@ -0,0 +1,237 @@
+	if err != nil {
+		return err
+	}
+	*deployment.Spec.Replicas = replicas
+	_, err = scaler.k.clientset.ExtensionsClient.Deployments(scaler.k.namespace).Update(deployment)
+
+	return err
+}
+
+// ReplicationControllerScaler implements ScalerInterface.
+type ReplicationControllerScaler struct {
+	k *KubernetesClient
+}
+
+func (scaler *ReplicationControllerScaler) GetNumOfReplicas() (int32, error) {
+	rc, err := scaler.k.clientset.CoreClient.ReplicationControllers(scaler.k.namespace).Get(scaler.k.target.targetName)
+	if err != nil {
+		return 0, err
+	}
+	return int32(*rc.Spec.Replicas), nil
+}
+
+func (scaler *ReplicationControllerScaler) UpdateNumOfReplicas(replicas int32) error {
+	if replicas == 0 {
+		log.Fatalf("cannot update to 0 replicas")
+	}
+
+	rc, err := scaler.k.clientset.CoreClient.ReplicationControllers(scaler.k.namespace).Get(scaler.k.target.targetName)
+	if err != nil {
+		return err
+	}
+	*rc.Spec.Replicas = replicas
+	_, err = scaler.k.clientset.CoreClient.ReplicationControllers(scaler.k.namespace).Update(rc)
+
+	return err
+}
+
+// ReplicaSetScaler implements ScalerInterface.
+type ReplicaSetScaler struct {
+	k *KubernetesClient
+}
+
+func (scaler *ReplicaSetScaler) GetNumOfReplicas() (int32, error) {
+	rs, err := scaler.k.clientset.ExtensionsClient.ReplicaSets(scaler.k.namespace).Get(scaler.k.target.targetName)
+	if err != nil {
+		return 0, err
+	}
+	return int32(*rs.Spec.Replicas), nil
+}
+
+func (scaler *ReplicaSetScaler) UpdateNumOfReplicas(replicas int32) error {
+	if replicas == 0 {
+		log.Fatalf("cannot update to 0 replicas")
+	}
+
+	rs, err := scaler.k.clientset.ExtensionsClient.ReplicaSets(scaler.k.namespace).Get(scaler.k.target.targetName)
+	if err != nil {
+		return err
+	}
+	*rs.Spec.Replicas = replicas
+	_, err = scaler.k.clientset.ExtensionsClient.ReplicaSets(scaler.k.namespace).Update(rs)
+
+	return err
+}
+
+// FetchConfigMap - fetch the requested key from the configmap
+func (k *KubernetesClient) FetchConfigMap(namespace, configmap, key string) (configMap ConfigMap, err error) {
+	cm, err := k.clientset.CoreClient.ConfigMaps(k.namespace).Get(configmap)
+	if err != nil {
+		return ConfigMap{}, err
+	}
+	return ConfigMap{cm.Data[key], cm.ObjectMeta.ResourceVersion}, nil
+}
+
+// GetClusterStatus Count schedulable nodes and cores in our cluster
+func (k *KubernetesClient) GetClusterStatus() (clusterStatus ClusterStatus, err error) {
+	opt := api.ListOptions{Watch: false}
+
+	nodes, err := k.clientset.CoreClient.Nodes().List(opt)
+	if err != nil {
+		return ClusterStatus{}, err
+	}
+	clusterStatus.totalNodes = int32(len(nodes.Items))
+	var tc resource.Quantity
+	var sc resource.Quantity
+	for _, node := range nodes.Items {
+		tc.Add(node.Status.Capacity[apiv1.ResourceCPU])
+		if !node.Spec.Unschedulable {
+			clusterStatus.schedulableNodes++
+			sc.Add(node.Status.Capacity[apiv1.ResourceCPU])
+		}
+	}
+
+	tcInt64, tcOk := tc.AsInt64()
```

Comment:
Does this always work? The implementation of this seems a bit weird if you look at the code:

````
package main

import (
	"fmt"

	"k8s.io/kubernetes/pkg/api/resource"
)

func main() {
	x, _ := resource.ParseQuantity("1234")

	fmt.Printf("%+v\n", x)

	y, err := x.AsInt64()
	fmt.Println(y, " ", err)

	// Promotion to decimal representation
	x.ToDec()
	// Getting the quantity as an Int64 no longer works. Is this expected?
	y, err = x.AsInt64()
	fmt.Println(y, " ", err)
}
````